sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java

```diff
@@ -0,0 +1,179 @@
+  public static JdbcIO.PreparedStatementSetCaller getPreparedStatementSetCaller(
+      Schema.FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case ARRAY:
+        return (element, ps, i, fieldWithIndex) -> {
+          ps.setArray(
+              i + 1,
+              ps.getConnection()
+                  .createArrayOf(
+                      fieldType.getCollectionElementType().getTypeName().name(),
+                      element.getArray(fieldWithIndex.getIndex()).toArray()));
+        };
+      case LOGICAL_TYPE:
+        {
+          String logicalTypeName = fieldType.getLogicalType().getIdentifier();
+          JDBCType jdbcType = JDBCType.valueOf(logicalTypeName);
+          switch (jdbcType) {
+            case DATE:
+              return (element, ps, i, fieldWithIndex) -> {
+                ps.setDate(
+                    i + 1, new Date(element.getDateTime(fieldWithIndex.getIndex()).getMillis()));
+              };
+            case TIME:
+              return (element, ps, i, fieldWithIndex) -> {
+                ps.setTime(
+                    i + 1, new Time(element.getDateTime(fieldWithIndex.getIndex()).getMillis()));
+              };
+            case TIMESTAMP_WITH_TIMEZONE:
+              return (element, ps, i, fieldWithIndex) -> {
+                ps.setTimestamp(
+                    i + 1,
+                    new Timestamp(element.getDateTime(fieldWithIndex.getIndex()).getMillis()));
+              };
+            default:
+              return getPreparedStatementSetCaller(fieldType.getLogicalType().getBaseType());
+          }
+        }
+      default:
+        {
+          if (typeNamePsSetCallerMap.containsKey(fieldType.getTypeName())) {
+            return typeNamePsSetCallerMap.get(fieldType.getTypeName());
+          } else {
+            throw new RuntimeException(
+                fieldType.getTypeName().name()
+                    + " in schema is not supported while writing. Please provide statement and preparedStatementSetter");
+          }
+        }
+    }
+  }
+
+  private static JdbcIO.PreparedStatementSetCaller createBytesCaller() {
+    return (element, ps, i, fieldWithIndex) -> {
+      validateLogicalTypeLength(
+          fieldWithIndex.getField(), element.getBytes(fieldWithIndex.getIndex()).length);
+      ps.setBytes(i + 1, element.getBytes(fieldWithIndex.getIndex()));
+    };
+  }
+
+  private static JdbcIO.PreparedStatementSetCaller createStringCaller() {
+    return (element, ps, i, fieldWithIndex) -> {
+      validateLogicalTypeLength(
+          fieldWithIndex.getField(), element.getString(fieldWithIndex.getIndex()).length());
+      ps.setString(i + 1, element.getString(fieldWithIndex.getIndex()));
+    };
+  }
+
+  private static void validateLogicalTypeLength(Schema.Field field, Integer length) {
+    if (field.getType().getTypeName().isLogicalType()
+        && length >= Integer.parseInt(field.getType().getLogicalType().getArgument())) {
```

Comment:
I think logical type argument is not always set. If the argument is not set as an Integer, maybe we can further check base type?

Reply:
@amaliujia the case is if the string length of input field exceeds from the table max limit then exception should be raised, base type will still be same even if the length differs.

Reply:
If so,  could we check input field to see if it is bytes or string?

Reply:
Input fields are checked in the switch case then called createStringCaller() for string for bytes it's createBytesCaller()
https://github.com/JawadHyder/beam/blob/9bb7280a65a4ea506f36d8597aaee3500633ef5a/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java#L155
If only type check is enough then wouldn't it truncate the string/bytes if it exceeds the max limit from the table?

Reply:
O I see. Then that's good enough. I wasn't sure how this function is called. E.g. If this function is called to verify other types like Date, Time, etc.

Reply:
No these methods are called for string and bytes due to variable length. So one thing, if the argument is not set should we throw exception (the field details like argument is fetched from table schema so I don't think it can be null) or let it proceed with the insertion?

Reply:
Argument is an optional parameter. So I think if it is not set, we can proceed to insertion.

Reply:
So there are cases:
1. If argument is not set - bypass the check and proceed with insertion (can we use a warning message here as data might truncate due to length mismatch?)
2. If argument is set and it is not integer - do same as above
3. If argument is set and it is integer - throw exception if length mismatch

Let's finalize these cases or any other case you think might possible or we can raise separate jira for this and remove checks from here?

Reply:
Sounds good to all of these. It would be backward-compatible to allow bypassing checking in 1 and 2 such that existing types won't break if setting an argument.

Reply:
done